### PR TITLE
Fea 86 task 114

### PR DESCRIPTION
**Pull request by SHulce**

### CODE REVIEW

File: rest-server/README.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        "17": {
+            "name": "Kris Hoadley",
+            "password": "TheBestLikeNoOneEverWas"
+        }
+    }
+}
+```
+To get a user by id
+```
+http://localhost:8080/v1/aoi/api/users/12
```
Comment (kchoadley) [inline rest-server/README.md]: not aoi, cadg

File: rest-server/README.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        "12": {
+            "name": "Shawn Hulce",
+            "password": "ecluhwordpass!4"
+        }
+    }
+}
+```
+To get a user by partial name match, use a query like
+```
+http://localhost:8080/v1/aoi/api/users?name=kris
```
Comment (kchoadley) [inline rest-server/README.md]: not aoi, cadg

File: rest-server/src/controller/controller.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    }
+    return query_map;
+}
+json::value Controller::ResponseNotImpl(const http::method & method) {
+    auto response = json::value::object();
+    response["http_method"] = json::value::string(method);
+    response["serviceName"] = json::value::string("AOI Service");
+    return response;
+}
+}  // namespace aoi_rest
```
Comment (kchoadley) [inline rest-server/src/controller/controller.cpp]: not aoi_rest, cadg_rest

File: rest-server/src/controller/user_controller.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+std::string UserController::ParseUserID(const std::string& path) {
+    auto next_forward_slash = path.find("/", 1);
+    std::string id_as_string;
+    if (next_forward_slash == std::string::npos)  // no more slashes
+        id_as_string = path.substr(1);
+    else
+        id_as_string = path.substr(1, next_forward_slash - 1);
+    return id_as_string;
+}
+}  // namespace aoi_rest
```
Comment (kchoadley) [inline rest-server/src/controller/user_controller.cpp]: not aoi_rest, cadg_rest

File: rest-server/src/services/data_access_object.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            users__.erase(itr);
+            return true;
+        }
+    }
+    return false;
+}
+void DataAccessObject::AddUser(User user) {
+    users__.push_back(user);
+}
+}  // namespace aoi_rest
```
Comment (kchoadley) [inline rest-server/src/services/data_access_object.cpp]: not aoi_rest, cadg_rest

File: rest-server/src/controller/controller.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                query_map[name] = value;
+            }
+        }
+    }
+    return query_map;
+}
+json::value Controller::ResponseNotImpl(const http::method & method) {
+    auto response = json::value::object();
+    response["http_method"] = json::value::string(method);
+    response["serviceName"] = json::value::string("AOI Service");
```
Comment (kchoadley) [inline rest-server/src/controller/controller.cpp]: Not AOI, CADG

File: rest-server/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+find_package(Threads REQUIRED)
+find_package(OpenSSL 1.0.1 REQUIRED)
+set(OPENSSL_VERSION "1.0.1")
+
+message("-- configuring gcc options")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++17 -DBOOST_LOG_DYN_LINK")
+
+# source files
+add_executable(${PROJECT_NAME}  
+                                ./src/controller/user_controller.cpp
```
Comment (kchoadley) [inline rest-server/CMakeLists.txt]: These are out of order. All the controller directory files should be next to each other, everything should be in alphabetical order.

File: rest-server/src/rest_server.cpp [hunk 8]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2018   Vaniya Agrawal, Ross Arcemont, Kristofer Hoadley,
+//                  Shawn Hulce, Michael McCulley
+#include <iostream>
+#include <string>
+//#include "another_controller.hpp"
```
Comment (kchoadley) [inline rest-server/src/rest_server.cpp]: Just remove these, not needed

File: rest-server/src/rest_server.cpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+//#include "example_controller.hpp"
+#include "user_controller.hpp"
+#include "data_access_interface.hpp"
+#include "data_access_object.hpp"
+using cadg_rest::DataAccessInterface;
+using cadg_rest::DataAccessObject;
+//using cadg_rest::AnotherController;
```
Comment (kchoadley) [inline rest-server/src/rest_server.cpp]: Remove these

File: rest-server/src/rest_server.cpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+//using cadg_rest::ExampleController;
+using cadg_rest::UserController;
+
+int main(int argc, const char * argv[]) {
+    std::cout << "Starting cadg rest server" << std::endl;
+    DataAccessInterface* user_dao = &DataAccessObject::Instance();
+    //AnotherController another_controller;
```
Comment (kchoadley) [inline rest-server/src/rest_server.cpp]: Remove these

File: rest-server/src/rest_server.cpp [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    if (argc > 2)
+        server_address.append(argv[2]);
+    else
+        server_address.append("http://host_auto_ip4");
+    if (argc > 1)
+        server_address.append(argv[1]);
+    else
+        server_address.append(":8080");
+    server_address.append("/v1/cadg/api");
+    //another_controller.endpoint(server_address + "/another");
```
Comment (kchoadley) [inline rest-server/src/rest_server.cpp]: Remove these

File: rest-server/src/rest_server.cpp [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    //example_controller.endpoint(server_address + "/example");
+    user_controller.endpoint(server_address + "/users");
+    try {
+        //another_controller.Accept().wait();
```
Comment (kchoadley) [inline rest-server/src/rest_server.cpp]: Remove these

File: rest-server/src/rest_server.cpp [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        //example_controller.Accept().wait();
+        user_controller.Accept().wait();
+        std::cout << "Listening for requests at: "
+                << user_controller.endpoint() << std::endl;
+        std::cout << "Press ENTER to exit." << std::endl;
+        std::string line;
+        std::getline(std::cin, line);
+        //another_controller.Shutdown().wait();
```
Comment (kchoadley) [inline rest-server/src/rest_server.cpp]: Remove these

File: rest-server/README.md (hunk 1 above)
Comment (SHulce) [reply rest-server/README.md]: D'oh!

File: rest-server/src/controller/controller.cpp [hunk 14]
```diff
@@ -64,4 +64,4 @@ json::value Controller::ResponseNotImpl(const http::method & method) {
     response["serviceName"] = json::value::string("AOI Service");
```
Comment (kchoadley) [inline rest-server/src/controller/controller.cpp]: "AIO Service" should read "CADG Service"

**Pull request merged**